front/hooks/useEventSource.ts

```diff
@@ -1,67 +1,173 @@
-import { useEffect, useRef, useState } from "react";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+const RECONNECT_DELAY = 5000; // 5 seconds.
+
+/**
+ * Stable EventSource Manager
+ *
+ * This singleton object manages EventSource instances across the entire application.
+ * It provides methods to create, retrieve, and remove EventSource connections,
+ * ensuring that each unique connection is properly managed and can be accessed
+ * or closed as needed.
+ *
+ * Key features:
+ * - Maintains a single source of truth for all EventSource instances.
+ * - Prevents duplicate EventSource creations for the same unique identifier.
+ * - Allows for centralized management of EventSource lifecycle.
+ * - Improves performance by reusing existing connections when possible.
+ *
+ * Usage:
+ * - Create a new EventSource: stableEventSourceManager.create(url, uniqueId)
+ * - Retrieve an existing EventSource: stableEventSourceManager.get(uniqueId)
+ * - Close and remove an EventSource: stableEventSourceManager.remove(uniqueId)
+ *
+ * This manager is designed to be used in conjunction with the useEventSource hook
+ * to provide stable and efficient EventSource handling across React component lifecycles.
+ */
+const stableEventSourceManager = {
+  // Map to store active EventSource instances, keyed by unique identifiers
+  sources: new Map<string, EventSource>(),
```

Comment:
Can we have a weakmap here? This will prevent leaking these connections if we forget to call manager.remove().